crates/iota-data-ingestion-core/src/progress_store/file.rs

```diff
@@ -2,21 +2,108 @@
 // Modifications Copyright (c) 2024 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use std::path::PathBuf;
+use std::{io::SeekFrom, path::PathBuf};
 
 use async_trait::async_trait;
 use iota_types::messages_checkpoint::CheckpointSequenceNumber;
 use serde_json::{Number, Value};
+use tokio::{
+    fs::{File, OpenOptions},
+    io::{AsyncReadExt, AsyncSeekExt, AsyncWriteExt},
+};
 
-use crate::{IngestionError, progress_store::ProgressStore};
+use crate::{IngestionError, IngestionResult, progress_store::ProgressStore};
 
+/// Manages persistent progress information stored in a JSON file.
+///
+/// This struct encapsulates file operations for reading, writing, and
+/// synchronizing progress data to disk. It uses asynchronous I/O provided by
+/// [`tokio::fs`] for efficient operation within a Tokio runtime.
+///
+/// # Example
+/// ```
+/// use iota_data_ingestion_core::{FileProgressStore, ProgressStore};
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let mut store = FileProgressStore::new("progress.json").await.unwrap();
+///     store.save("task1".into(), 42).await.unwrap();
+///     let checkpoint = store.load("task1".into()).await.unwrap();
+///     assert_eq!(checkpoint, 42);
+///     # tokio::fs::remove_file("progress.json").await.unwrap();
+/// }
+/// ```
 pub struct FileProgressStore {
-    path: PathBuf,
+    /// The [`File`] handle used to interact with the progress file.
+    file: File,
 }
 
 impl FileProgressStore {
-    pub fn new(path: PathBuf) -> Self {
-        Self { path }
+    /// Creates a new `FileProgressStore` by opening or creating the file at the
+    /// specified path.
+    pub async fn new(path: impl Into<PathBuf>) -> IngestionResult<Self> {
+        Self::open_or_create_file(path.into())
+            .await
+            .map(|file| Self { file })
+    }
+
+    /// Open or create the file at the specified path.
+    async fn open_or_create_file(path: PathBuf) -> IngestionResult<File> {
+        OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .truncate(false)
+            .open(path)
+            .await
+            .map_err(Into::into)
```

Comment:
Nit: Use the `?` shorthand instead of `map_err`. In consistency with the rest of the API. Make for more concise code as well.
```suggestion
        Ok(OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .truncate(false)
            .open(path)
            .await?)
```